std/random.d

```diff
@@ -1361,17 +1620,126 @@ how excellent the source of entropy is.
     }
     else
     {
-        import core.thread : Thread, getpid, MonoTime;
-        static bool seeded;
-        static MinstdRand0 rand;
-        if (!seeded)
+        return cast(uint) .unpredictableSeed!ulong;
+    }
+}
+/// ditto
+template unpredictableSeed(UIntType)
+if (isUnsigned!UIntType)
+{
+    static if (!is(Unqual!UIntType == UIntType))
+        alias unpredictableSeed = .unpredictableSeed!(Unqual!UIntType);
+    else static if (is(UIntType == uint))
+        alias unpredictableSeed = .unpredictableSeed;
+    else
+    {
+        /// ditto
+        @property UIntType unpredictableSeed() @nogc nothrow @trusted
         {
-            uint threadID = cast(uint) cast(void*) Thread.getThis();
-            rand.seed((getpid() + threadID) ^ cast(uint) MonoTime.currTime.ticks);
-            seeded = true;
+            version (AnyARC4Random)
+            {
+                static if (UIntType.sizeof <= uint.sizeof)
+                {
+                    return cast(UIntType) arc4random();
+                }
+                else
+                {
+                    UIntType result = void;
+                    arc4random_buf(&result, UIntType.sizeof);
+                    return result;
+                }
+            }
+            else static if (!is(UIntType == ulong))
+            {
+                return cast(UIntType) unpredictableSeed!ulong;
+            }
+            else
+            {
+                /+
+                In the absence of a convenient arc4random-like interface use a
+                shared static SplitMix generator. Described by Guy L. Steele Jr.,
+                Doug Lea, and Christine H. Flood in "Fast Splittable Pseudorandom
+                Number Generators" (2014). As suggested in the paper use David
+                Stafford's superior parameterizations of the MurmurHash3 64-bit
+                finalizer from
+                http://zimbry.blogspot.com/2011/09/better-bit-mixing-improving-on.html.
+                +/
+                import core.atomic : atomicLoad, atomicOp, atomicStore, MemoryOrder;
+                enum bool has64BitAtomicOps = __traits(compiles,
+                {
+                    shared static ulong example;
+                    atomicOp!"+="(example, ulong.max);
+                });
+
+                static if (has64BitAtomicOps)
```

Comment:
It's so the code will compile on platforms that don't support atomic operations on 64 bit integers.

Reply:
OK.  But why write the code in these 2 different ways?  Why not just write the implementation that will work on systems without 64-bit atomic op support?

Reply:
Ah, I get it. Yes, you can call it an optimization to update with a single atomic operation rather than two atomic operations that can't be done in parallel, although in this case though I wrote the "optimized" version first then the 32-bit-only version second. I don't see a reason to delete the better branch when it is (to my eyes) clearer than the other branch.

>a fallback mechanism for something that is not optimized for speed in other respects

Although it is not the most important consideration speed is considered throughout the design. The reason I chose to keep SplitMix as the global seeder even after discarding the idea of "splitting" is that its state can be updated using solely atomic operations rather than requiring a mutex. The reason I don't [open and close /dev/urandom on every `unpredictableSeed` call](https://github.com/dlang/phobos/pull/5230#issuecomment-335267797) and instead use /dev/urandom to initialize a global seeder is speed.  I wrote this with the awareness that user code out in the wild might be [using `unpredictableSeed` incorrectly](https://issues.dlang.org/show_bug.cgi?id=18217) and so (all else being equal) I would like to not make such code become catastrophically slow.

Reply:
> Yes, you can call it an optimization to update with a single atomic operation rather than two atomic operations that can't be done in parallel, although in this case though I wrote the "optimized" version first then the 32-bit-only version second. I don't see a reason to delete the better branch when it is (to my eyes) clearer than the other branch.

Well, I'd suggest an alternative consideration here.

This is a function that is not expected to be called many times in an application's lifetime.  So while speed is always nice, it's probably a secondary consideration to being clear, and simple, and easy to maintain.

Right now it looks like there is a series of design choices impacting on one another:

* there's a choice to use `shared` state for the fallback mechanism

* that means that you need to either use mutexes or atomic ops

* that means that you need to care about whether the system has support for 64-bit atomic ops

* that means that you have to supply multiple different implementations

What if, instead, you simply used a single, thread-local `static` splitmix instance?  It sounds like this would get rid of any need to care about atomic operations, and I don't think there would be any meaningful cost in terms of speed.

It would make the code _much_ simpler to implement, and much easier to understand and modify in future.  And it would fit nicely with separating out the splitmix implementation into a `struct SplitMixEngine`.

Reply:
>What if, instead, you simply used a single, thread-local static splitmix instance

That's an option. We still need to seed it.

I could also just delete the branch for architectures that don't have 64-bit atomic ops. Right now `core.atomic.has64BitCAS` is true for DMD on every platform it supports. Same for LDC.

Reply:
> That's an option. We still need to seed it.

This is _one_ of the reasons why I strongly recommend we separate out the concerns of this PR.

Right now we have a simple (and imperfect) seeding mechanism, which is the one using pid and tid and current time.  It's obviously nice to replace this.  But the introduction of splitmix doesn't need to depend on that.

So, why not start by just implementing splitmix -- in a cleanly separated way -- and updating `unpredictableSeed` to use it, again in the simplest possible way?

Then, as a separate PR, we can add support for the different system entropy sources, and update `unpredictableSeed` to make use of them -- either directly (as sources of the returned value) or indirectly (to generate the seed for splitmix).

Reply:
>But the introduction of splitmix doesn't need to depend on that.

This SplitMix is initialized with two 64-bit values and our current seeding method produces a single 32-bit value, so minimally we would need to also add a stopgap seeding mechanism that we would immediately replace with the real seeding mechanism in a subsequent pull request. I think that is bad.

Reply:
The motivation for introducing the global PRNG was to drastically improve the fallback case. In the case where we can't get system entropy at all, we aren't reduced to initializing thread-local PRNGs with very similar tid+pid+time seeds. Using the global PRNG directly was the next logical step from there.